easybuild/tools/job/__init__.py

```diff
@@ -0,0 +1,122 @@
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://vscentrum.be/nl/en),
+# the Hercules foundation (http://www.herculesstichting.be/in_English)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# http://github.com/hpcugent/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""Abstract interface for submitting jobs and related utilities."""
+
+
+from abc import ABCMeta, abstractmethod
+
+from vsc.utils.missing import get_subclasses
+
+from easybuild.tools.config import get_job_backend
+from easybuild.tools.config import PREFERRED_JOB_BACKENDS
+from easybuild.tools.utilities import import_available_modules
+
+
+class JobServer(object):
+    __metaclass__ = ABCMeta
+
+    USABLE = False
+
+    @abstractmethod
+    def begin(self):
+        """
+        Start a bulk job submission.
+
+        Jobs may be queued and only actually submitted when `commit()`
+        is called.
+        """
+        pass
+
+    @abstractmethod
+    def make_job(self, script, name, env_vars=None, hours=None, cores=None):
+        """
+        Create and return a `Job` object with the given parameters.
+
+        See the `Job`:class: constructor for an explanation of what
+        the arguments are.
+        """
+        pass
+
+    @abstractmethod
+    def submit(self, job, after=frozenset()):
+        """
+        Submit a job to the batch-queueing system.
+
+        If second optional argument `after` is given, it must be a
+        sequence of jobs that must be successfully terminated before
+        the new job can run.
+
+        Note that actual submission may be delayed until `commit()` is
+        called.
+        """
+        pass
+
+    @abstractmethod
+    def commit(self):
+        """
+        End a bulk job submission.
+
+        Releases any jobs that were possibly queued since the last
+        `begin()` call.
+
+        No more job submissions should be attempted after `commit()`
+        has been called, until a `begin()` is invoked again.
+        """
+        pass
+
+
+def avail_job_backends(check_usable=True):
+    """
+    Return all known job execution backends.
+    """
+    import_available_modules('easybuild.tools.job')
+    class_dict = dict([(x.__name__, x)
+                       for x in get_subclasses(JobServer)
+                       if (x.USABLE or not check_usable)])
+    return class_dict
```

Comment:
@stdweird I see your concern, but I do not see how `avail_job_backends()` could help here: what should it do if there are "junk" strings in `PREFERRED_JOB_BACKENDS`?  Error out and stop EB? But then why should it stop if the typo is not in the preferred/most significant backend?

IMHO, the place to verify the content of `PREFERRED_JOB_BACKENDS` would be to have an assertion in `easybuil/tools/config.py`, right after the definition.  This way, any typo would be caught by pre-release testing.  For example::

    PREFERRED_JOB_BACKENDS = ('Pbs', 'GC3Pie')
    def _verify_preferred_job_backends():
        from easybuild.tools.job import avail_job_backends
        available = avail_job_backends(check_usable=False)
        return set(PREFERRED_JOB_BACKENDS) < set([ cls.__name__ for cls in available ])
    assert _verify_preferred_job_backends()

Another possibility is to use class names instead of strings; but that requires that `config.py` imports all available job backend modules::

    from easybuild.tools.job.pbs import Pbs
    from easybuild.tools.job.gc3pie import GC3Pie

    # ...

    PREFERRED_JOB_BACKENDS = (Pbs, GC3Pie)

then `get_job_backend()` needs to be changed to translate the class name to the actual class.